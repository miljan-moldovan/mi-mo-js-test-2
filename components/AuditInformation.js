import React from 'react';
import { View, Text, StyleSheet, ActivityIndicator } from 'react-native';
import moment from 'moment';
import { reverse, slice } from 'lodash';

import SalonTouchableOpacity from './SalonTouchableOpacity';

const AuditType = {
  NULL: -1,
  Normal: 0,
  Change: 1,
  Cancel: 2,
  ReviewWeb: 10,
};

const DateTime = {
  time: 'HH:mm:ss',
  timeOld: 'HH:mm',
  displayTime: 'hh:mm A',
  date: 'YYYY-MM-DD',
  dateTime: 'YYYY-MM-DDT00:00:00',
  serverDateTime: 'YYYY-MM-DDTHH:mm:ss',
};

const styles = StyleSheet.create({
  panelInfo: {
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    width: '92%',
    backgroundColor: '#F1F1F1',
    flex: 1,
    marginBottom: 14,
    paddingVertical: 15,
  },
  panelInfoLine: {
    flex: 1,
    paddingBottom: 5,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
  panelInfoTitle: {
    marginTop: 5,
    color: '#000000',
    fontSize: 14,
    fontFamily: 'Roboto',
    backgroundColor: 'transparent',
  },
  panelInfoText: {
    color: '#3F3F3F',
    fontSize: 11,
    fontFamily: 'Roboto',
    backgroundColor: 'transparent',
  },
  panelInfoDate: {
    color: '#000000',
    fontSize: 11,
    fontFamily: 'Roboto',
    backgroundColor: 'transparent',
  },
  panelInfoShowMoreText: {
    color: '#115ECD',
    fontSize: 10,
    fontFamily: 'Roboto',
    backgroundColor: 'transparent',
    textAlign: 'center',
  },
  panelInfoShowMore: {
    borderColor: '#CACBCF',
    borderWidth: 1,
    borderRadius: 10,
    backgroundColor: '#FFFFFF',
    width: 100,
    position: 'absolute',
    bottom: 12,
  },
});

export default class AuditInformation extends React.Component {
  constructor(props) {
    super(props);
    const prepareData = this.prepareAudit(props.audit);
    this.state = {
      isOpen: false,
      audit: prepareData.audits,
      isBlockTime: prepareData.isBlockTime,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.audit) {
      const prepareData = this.prepareAudit(nextProps.audit);

      this.setState({
        audit: prepareData.audits,
        isBlockTime: prepareData.isBlockTime,
      });
    }
  }

  getAuditType = (auditType) => {
    let type = '';
    switch (auditType) {
      case AuditType.Normal:
        type = 'BOOKED';
        break;
      case AuditType.Change:
        type = 'MODIFIED';
        break;
      case AuditType.Cancel:
        type = 'CANCEL';
        break;
      case AuditType.ReviewWeb:
        type = 'REVIEW-WEB';
        break;
      default:
        type = '--';
    }

    return type;
  };

  prepareAudit = (audits) => {
    let isBlockTime = false;
    const prepareAudits = audits.map((audit) => {
      let oneAudit = null;
      if (!audit.service) {
        isBlockTime = true;
        oneAudit = {
          appointmentDate: audit.scheduleBlockDate || '',
          appointmentStartTime: audit.scheduleBlockStartTime || '',
          appointmentEndTime: audit.scheduleBlockEndTime || '',
          auditType: audit.auditType,
          auditDateTime: audit.auditDateTime || '',
          auditEmployee: audit.auditEmployee ? audit.auditEmployee.name : '',
        };
      } else {
        oneAudit = {
          appointmentDate: audit.appointmentDate || '',
          appointmentStartTime: audit.appointmentStartTime || '',
          appointmentEndTime: audit.appointmentEndTime || '',
          provider: audit.provider ? audit.provider.name : '',
          service: audit.service ? audit.service.name : '',
          auditType: audit.auditType,
          auditDateTime: audit.auditDateTime || '',
          auditEmployee: audit.auditEmployee ? audit.auditEmployee.name : '',
        };
      }
      return oneAudit;
    });
    return { audits: reverse(prepareAudits), isBlockTime };
  };

  formatDate = date => moment(date, DateTime.date).format('MM/DD/YYYY');

  formatTime = time => moment(time, DateTime.time).format(DateTime.displayTime);

  formatTimeWithMinutes = time =>
    moment(time, DateTime.serverDateTime).format(DateTime.displayTime);

  formatEmployeeName = (employee) => {
    if (!employee) {
      return '';
    }
    const splitName = employee.split(' ');
    switch (splitName.length) {
      case 0:
        return '';
      case 1:
        return splitName[0];
      case 2:
        return `${splitName[0]} ${splitName[1][0]}`;
      default:
        return `${splitName[0]} ${splitName[2][0]}`;
    }
  };

  renderAuditInfo = (audit, isBlockTime, isOpen) => {
    let auditToRender = audit;
    if (!isOpen && !this.props.isLoading && audit && audit.length > 1) {
      auditToRender = slice(auditToRender, 0, 1);
    }
    return auditToRender.map(item => (
      <View key={Math.random()} style={[styles.panelInfoLine, { paddingTop: 5 }]}>
        <Text style={styles.panelInfoTitle}>{this.getAuditType(item.auditType)}</Text>
        {isBlockTime
          ? [
            <Text style={styles.panelInfoText}>
                at {this.formatDate(item.auditDateTime)},{' '}
              {item.auditDateTime ? ` ${this.formatTimeWithMinutes(item.auditDateTime)} ` : ''}
            </Text>,
            <Text style={styles.panelInfoDate}>
              {this.formatEmployeeName(item.auditEmployee)}
            </Text>,
            ]
          : [
            <Text style={styles.panelInfoText}>
              {item.service || ''} with {this.formatEmployeeName(item.provider)} on{' '}
              {this.formatDate(item.appointmentDate)} at{' '}
              {this.formatTime(item.appointmentStartTime)}
            </Text>,
            <Text style={styles.panelInfoDate}>
                by {this.formatEmployeeName(item.auditEmployee)} on{' '}
              {item.auditDateTime ? this.formatDate(item.auditDateTime) : ''}
            </Text>,
            ]}
      </View>
    ));
  };

  render() {
    const { isBlockTime, isOpen } = this.state;
    const { audit } = this.state;
    return this.props.isLoading ? (
      <View
        style={{
          flex: 1,
          alignItems: 'center',
          justifyContent: 'center',
        }}
      >
        <ActivityIndicator />
      </View>
    ) : (
      [
        <View style={[styles.panelInfo, isOpen ? {} : { maxHeight: 230 }]}>
          {this.renderAuditInfo(audit, isBlockTime, isOpen)}
        </View>,
        <SalonTouchableOpacity
          style={styles.panelInfoShowMore}
          onPress={() => {
            this.setState({ isOpen: !this.state.isOpen });
          }}
        >
          <Text style={styles.panelInfoShowMoreText}>
            {this.state.isOpen ? 'SHOW LESS' : 'SHOW MORE'}
          </Text>
        </SalonTouchableOpacity>,
      ]
    );
  }
}
