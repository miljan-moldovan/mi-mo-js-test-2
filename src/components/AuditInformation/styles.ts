import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  spinner: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  panelInfo: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'stretch',
    width: '92%',
    backgroundColor: '#F1F1F1',
    padding: 15,
  },
  panelInfoIsOpened: {
    maxHeight: 230,
    minHeight: 55,
  },
  panelInfoLine: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
  panelInfoShowMoreText: {
    color: '#115ECD',
    fontSize: 10,
    fontFamily: 'Roboto',
    backgroundColor: 'transparent',
    textAlign: 'center',
  },
  panelInfoShowMore: {
    borderColor: '#CACBCF',
    borderWidth: 1,
    borderRadius: 10,
    backgroundColor: '#FFFFFF',
    width: 100,
    position: 'absolute',
    bottom: 6,
  },
});

export default styles;
