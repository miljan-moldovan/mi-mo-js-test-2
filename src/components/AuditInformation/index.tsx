// tslint:disable:max-line-length
import * as React from 'react';
import { View, Text, ActivityIndicator } from 'react-native';
import { reverse, slice } from 'lodash';
import SalonTouchableOpacity from '../SalonTouchableOpacity';
import styles from './styles';
import AuditInfoItem from '../AuditItem';

export default class AuditInformation extends React.Component<any, any> {
  constructor(props) {
    super(props);
    const { audits, isBlockTime } = this.prepareAudit(props.audit);
    this.state = {
      isOpen: false,
      audit: audits,
      isBlockTime,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.audit) {
      const { audits, isBlockTime } = this.prepareAudit(nextProps.audit);

      this.setState({
        audit: audits,
        isBlockTime,
      });
    }
  }

  prepareAudit = audits => {
    let isBlockTime = false;
    const prepareAudits = audits.map(audit => {
      if (!audit.service) {
        isBlockTime = true;
        return {
          appointmentDate: audit.scheduleBlockDate || '',
          appointmentStartTime: audit.scheduleBlockStartTime || '',
          appointmentEndTime: audit.scheduleBlockEndTime || '',
          auditType: audit.auditType,
          auditDateTime: audit.auditDateTime || '',
          auditEmployee: audit.auditEmployee ? audit.auditEmployee.name : '',
        };
      }
      return {
        appointmentDate: audit.appointmentDate || '',
        appointmentStartTime: audit.appointmentStartTime || '',
        appointmentEndTime: audit.appointmentEndTime || '',
        provider: audit.provider ? audit.provider.name : '',
        service: audit.service ? audit.service.name : '',
        auditType: audit.auditType,
        auditDateTime: audit.auditDateTime || '',
        auditEmployee: audit.auditEmployee ? audit.auditEmployee.name : '',
      };
    });
    return { audits: reverse(prepareAudits), isBlockTime };
  };

  renderAuditInfo = (audit, isBlockTime, isOpen) => {
    let auditToRender = audit;
    if (!isOpen && !this.props.isLoading && audit && audit.length > 1) {
      auditToRender = slice(auditToRender, 0, 1);
    }
    return auditToRender.map(item => (
      <AuditInfoItem key={Math.random()} singleAudit={item} isBlockTime={isBlockTime} />
    ));
  };

  render() {
    const { isBlockTime, isOpen, audit } = this.state;

    return this.props.isLoading ?
      (
        <View style={styles.spinner}>
          <ActivityIndicator />
        </View>
      )
    :
      (
        <React.Fragment>
          <View style={[styles.panelInfo, isOpen ? {} : { maxHeight: 230, minHeight: 55 }]}>
            {this.renderAuditInfo(audit, isBlockTime, isOpen)}
          </View>
          {audit.length > 1 &&
            (
              <SalonTouchableOpacity
                style={styles.panelInfoShowMore}
                onPress={() => {
                  this.setState({ isOpen: !this.state.isOpen });
                }}
              >
                <Text style={styles.panelInfoShowMoreText}>
                  {this.state.isOpen ? 'SHOW LESS' : 'SHOW MORE'}
                </Text>
              </SalonTouchableOpacity>
            )
          }
        </React.Fragment>
    );
  }
}
