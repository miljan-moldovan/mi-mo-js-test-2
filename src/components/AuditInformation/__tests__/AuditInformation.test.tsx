import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { ActivityIndicator } from 'react-native';
import AuditInformation from '../index';
import sinon from 'sinon';

const auditMock = [
  {
    id: 1719035,
    auditType: 0,
    auditDateTime: '2018-12-18T10:07:21',
    auditEmployee: {
      middleName: null,
      lastName: 'Zona',
      fullName: 'Frank Zona',
      type: 1,
      isTerminated: false,
      inAppointmentBook: true,
      name: 'Frank',
      id: 6,
      updateStamp: 1544777306892251,
      isDeleted: false,
    },
    appointmentId: 762238,
    appointmentDate: '2018-12-18T00:00:00',
    appointmentStartTime: '12:15:00',
    appointmentEndTime: '14:00:00',
    provider: {
      name: 'Bridget',
      id: 389,
      updateStamp: 1545136458131326,
      isDeleted: false,
    },
    service: {
      name: '1 Formula Foil',
      id: 30,
      updateStamp: 1529495311771391,
      isDeleted: false,
    },
  },
  {
    id: 1719036,
    auditType: 1,
    auditDateTime: '2018-12-18T10:07:29',
    auditEmployee: {
      middleName: null,
      lastName: 'Mini',
      fullName: 'Bridget Mini',
      type: 1,
      isTerminated: false,
      inAppointmentBook: true,
      name: 'Bridget',
      id: 389,
      updateStamp: 1545136458131326,
      isDeleted: false,
    },
    appointmentId: 762238,
    appointmentDate: '2018-12-18T00:00:00',
    appointmentStartTime: '12:15:00',
    appointmentEndTime: '14:00:00',
    provider: {
      name: 'Bridget',
      id: 389,
      updateStamp: 1545136458131326,
      isDeleted: false,
    },
    service: {
      name: '1 Formula Foil',
      id: 30,
      updateStamp: 1529495311771391,
      isDeleted: false,
    },
  },
  {
    id: 1719037,
    auditType: 1,
    auditDateTime: '2018-12-18T10:15:42',
    auditEmployee: {
      middleName: null,
      lastName: 'Zona',
      fullName: 'Frank Zona',
      type: 1,
      isTerminated: false,
      inAppointmentBook: true,
      name: 'Frank',
      id: 6,
      updateStamp: 1544777306892251,
      isDeleted: false,
    },
    appointmentId: 762238,
    appointmentDate: '2018-12-18T00:00:00',
    appointmentStartTime: '12:15:00',
    appointmentEndTime: '13:15:00',
    provider: {
      name: 'Bridget',
      id: 389,
      updateStamp: 1545136458131326,
      isDeleted: false,
    },
    service: {
      name: '60 Minutes',
      id: 14,
      updateStamp: 1544024336731581,
      isDeleted: false,
    },
  },
];

describe('<AuditInformation />', () => {
  let defaultProps;
  let mountedComponent;

  // just for remove error with depend using react-dom for render. Need it before use adapter for RN
  const origConsole = console.error;
  const mountComponent = (props = defaultProps) => {
    if (!mountedComponent) {
      mountedComponent = mount(<AuditInformation {...props} />);
    }
    return mountedComponent;
  };

  beforeEach(() => {
    // just for remove error with depend using react-dom for render. Need it before use adapter for RN
    console.error = () => {};
    defaultProps = {
      isLoading: false,
      audit: [],
    };
    mountedComponent = undefined;
  });
  afterEach(() => {
    // just for remove error with depend using react-dom for render. Need it before use adapter for RN
    console.error = origConsole;
  });

  it('renders correctly', () => {
    expect(mountComponent()).toMatchSnapshot();
  });

  describe('Should render ActivityIndicator based on `isLoading` prop', () => {
    it('Should not render spinner when `isLoading` prop equals to `false`', () => {
      const spinner = mountComponent().contains(<ActivityIndicator />);
      expect(spinner).toBe(false);
    });
    it('Should render spinner when `isLoading` prop equals to `true`', () => {
      defaultProps.isLoading = true;
      const spinner = mountComponent().contains(<ActivityIndicator />);
      expect(spinner).toBe(true);
    });
  });

  describe('Should render a correct number of AuditInfoItem`s', () => {
    beforeEach(() => {
      defaultProps.audit = auditMock;
    });
    it('Should render a single item when `isOpen` is `false`', () => {
      const total = mountComponent()
        .find('AuditInfoItem')
        .reduce((amount, n) => (n ? amount + 1 : amount), 0);
      expect(total).toBe(1);
    });
    it('When `isOpen` is `true` number of AuditInfoItem`s should match audits.length', () => {
      const total = mountComponent()
        .setState({ isOpen: true })
        .find('AuditInfoItem')
        .reduce((amount, n) => (n ? amount + 1 : amount), 0);
      expect(total).toBe(auditMock.length);
    });
  });

  describe('Should correctly render `show more/less` button', () => {
    it('Should render the button when there`re several AuditItems', () => {
      defaultProps.audit = auditMock;
      const buttonIsRendered = mountComponent().exists('SalonTouchableOpacity(TouchableOpacity)');
      expect(buttonIsRendered).toBe(true);
    });
    it('Should NOT render the button when there`s a single AuditItem', () => {
      defaultProps.audit = auditMock.slice(0, 1);
      const buttonIsRendered = mountComponent().exists('SalonTouchableOpacity(TouchableOpacity)');
      expect(buttonIsRendered).toBe(false);
    });
  });

  describe('The `show more/less` button should properly change state', () => {
    beforeEach(() => {
      defaultProps.audit = auditMock;
    });

    it('Click on the button should change `isOpen`', () => {
      const component = mountComponent();
      expect(component.state().isOpen).toBe(false);

      component
        .find('SalonTouchableOpacity(TouchableOpacity)')
        .props()
        .onPress();
      expect(component.state().isOpen).toBe(true);
    });
  });

  describe('Should properly prepare audit data for render', () => {
    let spy;
    beforeEach(() => {
      defaultProps.audit = auditMock;
      spy = sinon.spy(AuditInformation.prototype, 'componentWillReceiveProps');
    });
    afterEach(() => {
      spy.restore();
    });

    it('Should run `prepareAudit` function and change state when received new audit props', () => {
      const component = mountComponent();

      expect(spy).toHaveProperty('callCount', 0);
      expect(component.state().isBlockTime).toBe(false);

      component.setProps({ audit: [{}, {}, {}] });

      expect(spy).toHaveProperty('callCount', 1);
      expect(component.state().isBlockTime).toBe(true);
    });
    it('Should NOT change state when received same audit props', () => {
      const component = mountComponent();

      expect(spy).toHaveProperty('callCount', 0);
      expect(component.state().isBlockTime).toBe(false);
      const firstState = component.state();

      component.setProps({ audit: auditMock });

      expect(spy).toHaveProperty('callCount', 1);
      expect(component.state().isBlockTime).toBe(false);
      expect(component.state()).toEqual(firstState);
    });
    it('Should NOT change state when DID NOT receive audit props', () => {
      const component = mountComponent();
      const firstAuditState = component.state('audit');

      component.setProps({ audit: null });

      expect(component.state('audit')).toBe(firstAuditState);
    });
  });

  describe('`prepareAudit` function should properly parse props', () => {
    beforeEach(() => {
      defaultProps.audit = auditMock;
    });

    it('Should properly parse audit with a lot of undefined values', () => {
      const emptyAuditParsed = {
        appointmentDate: '',
        appointmentStartTime: '',
        appointmentEndTime: '',
        provider: '',
        service: 'someMockService',
        auditType: 'someMockType',
        auditDateTime: '',
        auditEmployee: '',
      };

      const emptyAudit = {
        service: {
          name: 'someMockService',
        },
        auditType: 'someMockType',
      };

      const component = mountComponent({ audit: [emptyAudit] });

      expect(component.state('audit')).toEqual([emptyAuditParsed]);
      expect(component.state('isBlockTime')).toBe(false);
    });
    it('Should properly parse a completely empty object', () => {
      const undefinedAuditParsed = {
        appointmentDate: '',
        appointmentStartTime: '',
        appointmentEndTime: '',
        auditType: undefined,
        auditDateTime: '',
        auditEmployee: '',
        id: undefined,
      };

      const component = mountComponent({ audit: [{}] });

      expect(component.state('audit')).toEqual([undefinedAuditParsed]);
      expect(component.state('isBlockTime')).toBe(true);
    });
  });
});
