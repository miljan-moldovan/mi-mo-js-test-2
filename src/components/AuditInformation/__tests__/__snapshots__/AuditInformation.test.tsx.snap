// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuditInformation /> renders correctly 1`] = `
<AuditInformation
  audit={Array []}
  isLoading={false}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "backgroundColor": "#F1F1F1",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "padding": 15,
          "width": "92%",
        },
        Object {
          "maxHeight": 230,
          "minHeight": 55,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "stretch",
            "backgroundColor": "#F1F1F1",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 15,
            "width": "92%",
          },
          Object {
            "maxHeight": 230,
            "minHeight": 55,
          },
        ]
      }
    />
  </View>
</AuditInformation>
`;
